src/mapping/query/function/parse_duration.rs

```diff
@@ -0,0 +1,176 @@
+use super::prelude::*;
+use lazy_static::lazy_static;
+use regex::Regex;
+use rust_decimal::{prelude::ToPrimitive, Decimal};
+use std::collections::HashMap;
+use std::str::FromStr;
+
+lazy_static! {
+    static ref RE: Regex =
+        Regex::new(r"(?i)\A(?P<value>[0-9]*\.?[0-9]+)\s?(?P<unit>[a-z]{1,2})\z").unwrap();
+    static ref UNITS: HashMap<String, Decimal> = vec![
+        ("ns", Decimal::new(1, 9)),
+        ("µs", Decimal::new(1, 6)),
+        ("ms", Decimal::new(1, 3)),
+        ("cs", Decimal::new(1, 2)),
+        ("ds", Decimal::new(1, 1)),
+        ("s", Decimal::new(1, 0)),
+        ("m", Decimal::new(60, 0)),
+        ("h", Decimal::new(3_600, 0)),
+        ("d", Decimal::new(86_400, 0)),
+    ]
+    .into_iter()
+    .map(|(k, v)| (k.to_owned(), v))
+    .collect();
+}
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ParseDurationFn {
+    query: Box<dyn Function>,
+    format: Box<dyn Function>,
+}
+
+impl ParseDurationFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(query: Box<dyn Function>, format: &str) -> Self {
+        let format = Box::new(Literal::from(Value::from(format)));
+
+        Self { query, format }
+    }
+}
+
+impl Function for ParseDurationFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let value = {
+            let bytes = required!(ctx, self.query, Value::Bytes(v) => v);
+            String::from_utf8_lossy(&bytes).into_owned()
+        };
+
+        let fmt_decimal = {
+            let bytes = required!(ctx, self.format, Value::Bytes(v) => v);
+            let format = String::from_utf8_lossy(&bytes).into_owned();
+
+            UNITS
+                .get(&format)
+                .ok_or(format!("unknown format unit: '{}'", format))?
+        };
+
+        let captures = RE
+            .captures(&value)
+            .ok_or(format!("unable to parse duration: '{}'", value))?;
+
+        let value = Decimal::from_str(&captures["value"]).map_err(|e| e.to_string())?;
+        let decimal = UNITS
+            .get(&captures["unit"])
+            .ok_or(format!("unknown duration unit: '{}'", &captures["unit"]))?;
+
+        let number = value * decimal / fmt_decimal;
+        let number = number
+            .to_f64()
+            .ok_or(format!("unable to format duration: '{}'", number))?;
+
+        Ok(Value::from(number))
+    }
+
+    fn parameters() -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "format",
```

Comment:
@binarylogic suggested `unit`, does that work? I'm happy to use whatever. I admittedly had some trouble finding the right terminologies for the different parts of this function.

Reply:
I'm still uncomfortable with `unit`, as it could mean either the implied (or expected) input units if the source text is missing units, or the output unit conversion. I'd suggest `output_unit` or `output_conversion`, or simply `output`.

Reply:
Thanks @bruceg. I went with `output` for now.